### PR TITLE
[Darwin] Enable taking assertion on MTRDeviceController

### PR DESCRIPTION
**Pull request by anush-apple**
- Added ability to take assertion on MTRDeviceController to keep it running until all assertions are removed
- A request to shutdown will take place only if there are no assertions are present

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 1]
```diff
@@ -142,6 +142,10 @@ - (instancetype)initForSubclasses
         // nothing, as superclass of MTRDeviceController is NSObject
     }
     _underlyingDeviceMapLock = OS_UNFAIR_LOCK_INIT;
+
+    // Setup assertion counters
+    _keepRunningAssertionCounter = 0;
+    _shutdownPending = FALSE;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:148]: I think we mostly use YES/NO, not TRUE/FALSE, but either way.

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 2]
```diff
@@ -311,7 +320,53 @@ - (BOOL)isRunning
     return _cppCommissioner != nullptr;
 }
 
+- (NSUInteger)shutdownPrecheck
+{
+    __block NSUInteger assertionCount = 0;
+    dispatch_sync(_chipWorkQueue, ^{
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:326]: Do we actually have to do that on the work queue?  Or can we just do it under a lock?  That is, does the work queue use that value for anything?

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 3]
```diff
@@ -311,7 +320,53 @@ - (BOOL)isRunning
...
+    });
+    if (shouldShutdown) {
+        MTR_LOG("%@ All assertions removed and shutdown is pending, shutting down", self);
+        [self finalShutdown];
+    }
+}
+
 - (void)shutdown
+{
+    NSUInteger assertionCount = [self shutdownPrecheck];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:361]: So I think there is a race here: what if one thread calls shutdown while another is calling addRunAssertion?

We could not worry about that race, I guess, or we could make addRunAssertion fallible and have it fail if we are in a situation where shutdownPrecheck has already returned 0 (would have to set that state under the lock or on the matter queue in shutdownPrecheck).

File: src/darwin/Framework/CHIP/MTRDeviceController_Internal.h [hunk 4]
```diff
@@ -72,6 +72,10 @@ NS_ASSUME_NONNULL_BEGIN
 // (moved here so subclasses can initialize differently)
 @property (readwrite, retain) dispatch_queue_t chipWorkQueue;
 
+// Counters to track assertion status
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController_Internal.h:75]: Should document what synchronizes access to these, whether it's "must only touch on Matter queue" or some sort of locking.

File: src/darwin/Framework/CHIP/MTRDeviceControllerStartupParams.mm [hunk 5]
```diff
@@ -306,6 +306,29 @@ - (void)setOTAProviderDelegate:(id<MTROTAProviderDelegate>)otaProviderDelegate q
...
+}
+
++ (nullable NSNumber *)fabricIDFromNOC:(MTRCertificateDERBytes)noc
+{
+    NSNumber * fabricID = nil;
+    ExtractFabricIDFromNOC(noc, &fabricID);
+    return fabricID;
+}
+
++ (NSData *)publicKeyFromCertificate:(MTRCertificateDERBytes)certificate
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerStartupParams.mm:323]:
```suggestion
+ (nullable NSData *)publicKeyFromCertificate:(MTRCertificateDERBytes)certificate
```

File: src/darwin/Framework/CHIP/MTRDeviceControllerStartupParams_Internal.h [hunk 6]
```diff
@@ -85,6 +85,10 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, strong, readonly, nullable) id<MTROTAProviderDelegate> otaProviderDelegate;
 @property (nonatomic, strong, readonly, nullable) dispatch_queue_t otaProviderDelegateQueue;
 
++ (nullable NSNumber *)nodeIDFromNOC:(MTRCertificateDERBytes)noc;
++ (nullable NSNumber *)fabricIDFromNOC:(MTRCertificateDERBytes)noc;
++ (NSData *)publicKeyFromCertificate:(MTRCertificateDERBytes)certificate;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceControllerStartupParams_Internal.h:90]:
```suggestion
+ (nullable NSData *)publicKeyFromCertificate:(MTRCertificateDERBytes)certificate;
```

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 7]
```diff
@@ -311,8 +331,68 @@ - (BOOL)isRunning
...
+- (BOOL)matchesPendingShutdownWithParams:(MTRDeviceControllerParameters *)parameters
+{
+    if (!parameters.operationalCertificate || !parameters.rootCertificate) {
+        return FALSE;
+    }
+    NSNumber * nodeID = [MTRDeviceControllerParameters nodeIDFromNOC:parameters.operationalCertificate];
+    NSNumber * fabricID = [MTRDeviceControllerParameters fabricIDFromNOC:parameters.operationalCertificate];
+    NSData * publicKey = [MTRDeviceControllerParameters publicKeyFromCertificate:parameters.rootCertificate];
+
+    std::lock_guard lock(_assertionLock);
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRDeviceController.mm:343]: Might be worth documenting why this is all reasonable if any of those bits above returned nil (basically they won't match us then).

**Review by bzbarsky-apple — APPROVED** [review]

File: src/darwin/Framework/CHIP/MTRDeviceController.mm [hunk 8]
```diff
@@ -609,6 +694,14 @@ - (BOOL)startup:(MTRDeviceControllerStartupParamsInternal *)startupParams
 
         self->_storedFabricIndex = fabricIdx;
         self->_storedCompressedFabricID = _cppCommissioner->GetCompressedFabricId();
+
+        chip::Crypto::P256PublicKey rootPublicKey;
+        if (_cppCommissioner->GetRootPublicKey(rootPublicKey) == CHIP_NO_ERROR) {
+            self.rootPublicKey = [NSData dataWithBytes:rootPublicKey.Bytes() length:rootPublicKey.Length()];
+            self.nodeID = @(_cppCommissioner->GetNodeId());
```
[the remark this answers is not in the document]
Comment (anush-apple) [reply src/darwin/Framework/CHIP/MTRDeviceController.mm:701]: Filed https://github.com/project-chip/connectedhomeip/issues/35183

File: src/darwin/Framework/CHIP/MTRDeviceController.mm (hunk 7 above)
Comment (anush-apple) [reply src/darwin/Framework/CHIP/MTRDeviceController.mm:343]: Filed https://github.com/project-chip/connectedhomeip/issues/35184

File: src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm [hunk 9]
```diff
@@ -1133,12 +1133,31 @@ - (void)operationalInstanceAdded:(chip::PeerId &)operationalID
...
 - (nullable MTRDeviceController *)initializeController:(MTRDeviceController *)controller
                                         withParameters:(MTRDeviceControllerParameters *)parameters
                                                  error:(NSError * __autoreleasing *)error
 {
     [self _assertCurrentQueueIsNotMatterQueue];
 
+    // If there is a controller already running with matching parameters, re-use it
+    MTRDeviceController * existingController = [self _findControllerMatchingParams:parameters];
+    if (existingController) {
+        return existingController;
```
[the remark this answers is not in the document]
Comment (anush-apple) [reply src/darwin/Framework/CHIP/MTRDeviceControllerFactory.mm:1158]: Filed https://github.com/project-chip/connectedhomeip/issues/35185

**Comment by bzbarsky-apple** [comment]
Fast-tracking platform-specific fix with platform owner review.

**Pull request merged**